src/EditorFeatures/CSharp/EventHookup/EventHookupCommandHandler_TabKeyCommand.cs

```diff
@@ -85,116 +78,179 @@ private void HandleTabWorker(ITextView textView, ITextBuffer subjectBuffer, Acti
+        // want to cancel the bg tasks kicked off as we need their values to actually emit the event.
+        EventHookupSessionManager.DismissExistingSessions(cancelBackgroundTasks: false);
+
+        var task = ExecuteCommandAsync(
+            args,
+            nextHandler,
+            applicableToSpan,
+            document,
+            eventNameTask,
+            caretPoint.Value);
+        task.CompletesAsyncOperation(token);
+
+        // At this point, we've taken control, so don't send the tab into the buffer.  But do dismiss the overall
+        // session.  We no longer need it.
+        return true;
+    }
 
-        string eventHandlerMethodName = null;
-        if (EventHookupSessionManager.CurrentSession.GetEventNameTask.Status == TaskStatus.RanToCompletion)
+    private async Task ExecuteCommandAsync(
+        TabKeyCommandArgs args,
+        Action nextHandler,
+        SnapshotSpan applicableToSpan,
+        Document document,
+        Task<string> getEventNameTask,
+        SnapshotPoint initialCaretPoint)
+    {
+        try
         {
-            eventHandlerMethodName = EventHookupSessionManager.CurrentSession.GetEventNameTask.WaitAndGetResult(cancellationToken);
+            _threadingContext.ThrowIfNotOnUIThread();
+            await ExecuteCommandWorkerAsync().ConfigureAwait(false);
         }
-
-        if (eventHandlerMethodName == null ||
-            EventHookupSessionManager.CurrentSession.TextView != textView)
+        catch (OperationCanceledException)
+        {
+        }
+        catch (Exception ex) when (FatalError.ReportAndCatch(ex))
         {
-            nextHandler();
-            EventHookupSessionManager.CancelAndDismissExistingSessions();
-            return;
         }
 
-        // This tab means we should generate the event handler method. Begin the code
-        // generation process.
-        GenerateAndAddEventHandler(textView, subjectBuffer, eventHandlerMethodName, nextHandler, cancellationToken);
-    }
-
-    private void GenerateAndAddEventHandler(ITextView textView, ITextBuffer subjectBuffer, string eventHandlerMethodName, Action nextHandler, CancellationToken cancellationToken)
-    {
-        _threadingContext.ThrowIfNotOnUIThread();
+        return;
 
-        using (Logger.LogBlock(FunctionId.EventHookup_Generate_Handler, cancellationToken))
+        async Task ExecuteCommandWorkerAsync()
         {
-            EventHookupSessionManager.CancelAndDismissExistingSessions();
+            var textView = args.TextView;
+            var subjectBuffer = args.SubjectBuffer;
+
+            var factory = document.Project.Solution.Workspace.Services.GetRequiredService<IBackgroundWorkIndicatorFactory>();
+            using var waitContext = factory.Create(
+                textView,
+                applicableToSpan,
+                CSharpEditorResources.Generating_event);
+
+            var cancellationToken = waitContext.UserCancellationToken;
```

Comment:
❓ Can we not move this code into the caller, and at that point have the ability to wrap the call to `ExecuteCommandWorkerAsync` in a `ReportAndCatchUnlessCancelled`?

Reply:
i tried a few forms and it ended up a little too cluttered.  I personally like separating otu the try/catch + async-flow into one area, and thne normal feature4 logic into another.